// views/index.js
const hyper = require('hyperhtml/index')
const {bind, wire} = hyper

const api = require('api')
const SimpleTable = require('components/simple_table')

module.exports = {
  CurrentView,
  SystemsView,
  NotFoundView,
}

function CurrentView () {
  return wire()`${getView()}`

  function getView () {
    switch (location.hash.slice(1)) {
      case '/': return IndexView()
      case '/systems': return SystemsView()
      default: return NotFoundView()
    }
  }
}

function SystemsView (props = {}) {
  return wire()`
    ${{
      placeholder: 'Loading',
      any: api.systems.list()
        .then(result => new SimpleTable({
          pagination: {
            itemsPerPage: 5,
          },
          items: result,
          columns: [
            {title: 'Title', property: 'title'},
            {title: 'Created', property: 'created'}
          ]
        }))
    }}`
}

function NotFoundView () {
  return wire()`
  <div>
    <h1>Nothing to see here</h1>
  </div>`
}

function IndexView () {
  return wire()`
  <div>
    <h1>Index</h1>
  </div>`
}
