// views/index.js
const hyper = require('hyperhtml/index')
const {bind, wire} = hyper

const defineRouter = require('sheet-router')
const qs = require('sheet-router/qs')

function update () {
  bind(document.body)`${require('components/Layout')()}`
}

const api = require('api')
const SimpleTable = require('components/simple_table')

const router = defineRouter({ default: '/404' }, [
  ['/', IndexView],
  ['/systems', SystemsView, [
    ['/:id', SystemsView]
  ]],
  ['/404', NotFoundView],
])

module.exports = {
  CurrentView,
  SystemsView,
  NotFoundView,
}

function CurrentView () {
  const search = location.hash.includes('?') ? qs(location.hash) : {}
  console.log(search)
  return wire()`${getView()}`

  function getView () {
    return router(location.hash.slice(1), search)
    // return router(location.pathname)
  }
}

function SystemsView (props = {}) {
  return wire()`
    ${{
      placeholder: 'Loading',
      any: api.systems.list()
        .then(result => new SimpleTable({
          pagination: {
            itemsPerPage: 5,
          },
          items: result,
          columns: [
            {title: 'Title', property: 'title', sortable: true},
            {title: 'Created', property: 'created', sortable: true}
          ]
        }))
    }}`
}

function NotFoundView () {
  return wire()`
  <div>
    <h1>Nothing to see here</h1>
  </div>`
}

function IndexView (params, search) {
  return wire()`
  <div>
    <h1>Index</h1>
    <pre>${JSON.stringify(params, null, 2)}</pre>
    <pre>${JSON.stringify(search, null, 2)}</pre>
  </div>`
}
