// components/simple_table/SimpleTable.js
const hyper = require('hyperhtml/index')
const {bind, wire} = hyper
const xtend = require('xtend')
const _ = require('lodash')
window._ = _

const SimpleTh = require('./SimpleTh')
const SimpleTr = require('./SimpleTr')

module.exports = class SimpleTable extends hyper.Component {
  get defaultState () {
    return {
      // all items that are currently rendered
      // type: Array<Any>
      items: [],

      // the items that have been selected in the UI
      // type: same as this.items
      selectedItems: [],

      appliedSorts: {},
    }
  }

  /**
    @param props {Object}
    @param props.initialState {Object}
    @param props.getData {Array<Filter> -> Promise<Array<Object>>}
  */
  constructor (props) {
    super()
    this.props = props
    const state = xtend(props.initialState, {
      items: props.items
    })
    this.setState(state)
  }

  updateOrder ({column, order}) {
  }

  sortBy ({column, order}) {
    const sorts = _(this.state.appliedSorts)
      .map((v, k) => [k, v])
      .filter(pair => !!pair[1])
      .valueOf()

    this.setState({
      appliedSorts: xtend(this.state.appliedSorts, {
        [column.property]: order
      }),
      items: _.orderBy(this.props.items, sorts.map(x => x[0]), sorts.map(x => x[1]))
    })
  }

  render() {
    const html = this.html.bind(this) // just to make Atom hilite the template
    return html`
      <div>
        <table class=table>
          <thead>
            <tr>
              ${this.props.columns.map(column => {
                return new SimpleTh(column, {
                  updateSort: (args) => this.updateOrder(args)
                })
              })}
            </tr>
          </thead>
          <tbody>
            ${this.state.items.map(item => {
              return new SimpleTr({ columns: this.props.columns, item})
            })}
          </tbody>
        </table>
      </div>`
  }
}
